// @popperjs/core/lib/popper.js
import { popperGenerator } from "./index.js";
import { viewport } from "./enums.js";
import eventListeners from "./modifiers/eventListeners.js";
import popperOffsets from "./modifiers/popperOffsets.js";
import detectOverflow from "./modifiers/detectOverflow.js";
import computeStyles from "./modifiers/computeStyles.js";
import applyStyles from "./modifiers/applyStyles.js";
import offset from "./modifiers/offset.js";
import flip from "./modifiers/flip.js";
import preventOverflow from "./modifiers/preventOverflow.js";
import arrow from "./modifiers/arrow.js";
var defaultModifiers = [eventListeners, popperOffsets, Object.assign({}, detectOverflow, {
  name: 'detectOverflow:preventOverflow'
}), Object.assign({}, detectOverflow, {
  name: 'detectOverflow:flip',
  options: {
    rootArea: viewport
  }
}), computeStyles, applyStyles, offset, flip, preventOverflow, arrow];
var createPopper = popperGenerator({
  defaultModifiers: defaultModifiers
}); // eslint-disable-next-line import/no-unused-modules

export { createPopper, popperGenerator };
